source/agora/test/NetworkManager.d

```diff
@@ -49,3 +49,114 @@ unittest
     nodes.all!(node => node.getBlockHeight() == 1)
         .retryFor(2.seconds, "Nodes should have same block height");
 }
+
+/// test behavior when a node sends bad block data
+unittest
+{
+    import agora.common.BanManager;
+    import agora.consensus.data.Block;
+    import agora.common.Config;
+    import agora.common.Metadata;
+    import agora.common.crypto.Key;
+    import core.time;
+    import geod24.LocalRest;
+    import std.algorithm;
+
+    /// node which returns bad blocks
+    static class BadNode : TestNode
+    {
+        ///
+        public this (Config config)
+        {
+            super(config);
+        }
+
+        /// return phony blocks
+        public override const(Block)[] getBlocksFrom (ulong block_height,
+            size_t max_blocks)
+        {
+            Block[] blocks;
+            Transaction[] last_tx;
+
+            auto gen_key = () @trusted { return getGenesisKeyPair(); }();
+
+            Block last_block;
+            // make 20 blocks which have an invalid previous hash
+            foreach (idx; 0 .. 20)
+            {
+                auto txs = () @trusted { return makeChainedTransactions(gen_key, last_tx, 1); }();
+                last_tx = txs;
+                auto block = makeNewBlock(last_block, txs);
+
+                // currently the only block validation the Ledger does is the height
+                block.header.height = idx + 10;
+                blocks ~= block;
+                last_block = block;
+            }
+
+            return blocks;
+        }
+
+        /// return block length as returned by function above
+        public override ulong getBlockHeight () { return 20; }
+    }
+
+    static class BadAPIManager : TestAPIManager
+    {
+        // base class uses a hashmap, can't depend on the order of nodes
+        public RemoteAPI!TestAPI[] nodes;
+
+        /// Initialize a new node
+        public override void createNewNode (PublicKey address, Config conf)
+        {
+            static bool first_node = true;
```

Comment:
`first_node == (this.nodes.length == 0)`

Reply:
It's a static initializer, that can't work.

Reply:
Ah nevermind.

Reply:
See what I mean? My brain..